### ISSUE TITLE
x/pkgsite: searching exact standard library package names

### ISSUE THREAD

**Issue by Deleplace** [golang/go#64358]
### What is the URL of the page with the issue?

https://pkg.go.dev/search?q=path&m=

### What is your user agent?

Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36

### What did you do?

Searched the exact full name of a package of the standard library : "path"

### What did you expect to see?

https://pkg.go.dev/path should be the first result

### What did you see instead?

there are many results, but https://pkg.go.dev/path is not in the list at all

# General problem

This is similar to #41369 but more specific: when exact search terms fail to show the relevant standard library package as 1st result.

The following rule should always hold:

**IF a package exists in the standard library with name _FOO_, AND a user is searching the exact term "FOO" in pkg.go.dev, THEN the standard library package _FOO_ must be the 1st result in the list**

I tried 284 such searches, here are my findings.

- When the full package name contains a "/" e.g. "compress/bzip", and the user is searching for the exact full name "compress/bzip", then no result list is displayed. Instead, the page is redirected to the correct stdlib package. **This works 100% of the time**
- When the full package name is a single word e.g. "time", and the user is searching for the exact word, then a result list is displayed. The correct stdlib package is most often the 1st result, with **5 annoying exceptions**
- When the full package name contains a "/" e.g. "compress/bzip", and the user is searching for the exact last word e.g. "bzip", then a result list is displayed. The correct stdlib package is often the 1st result, with **16 annoying exceptions**

The results of these 5 single-word searches are buggy:

Package | Search link | Result
-- | -- | --
path | [search path](https://pkg.go.dev/search?q=path&m=) | NOT FOUND
cmp	|  [search cmp](https://pkg.go.dev/search?q=cmp&m=) |  2nd pos
maps | [search maps](https://pkg.go.dev/search?q=maps&m=) | 8th pos
plugin | [search plugin](https://pkg.go.dev/search?q=plugin&m=) | 2nd pos
slices | [search slices](https://pkg.go.dev/search?q=slices&m=) | 4th pos

The results of these 16 single-word searches are buggy too, according to the rule:

Package | Search link | Result
-- | -- | --
runtime/coverage	| [search coverage](https://pkg.go.dev/search?q=coverage&m=)	| NOT FOUND
go/build/constraint	| [search constraint](https://pkg.go.dev/search?q=constraint&m=)	| 3rd pos
go/constant	| [search constant](https://pkg.go.dev/search?q=constant&m=)	| 3rd pos
go/doc/comment	| [search comment](https://pkg.go.dev/search?q=comment&m=)	| 6th pos
go/format	| [search format](https://pkg.go.dev/search?q=format&m=)	| 2nd pos
go/importer	| [search importer](https://pkg.go.dev/search?q=importer&m=)	| 2nd pos
go/types	| [search types](https://pkg.go.dev/search?q=types&m=)	| 4th pos
image/color	| [search color](https://pkg.go.dev/search?q=color&m=)	| 2nd pos
net/rpc	| [search rpc](https://pkg.go.dev/search?q=rpc&m=)	| 3rd pos
regexp/syntax	| [search syntax](https://pkg.go.dev/search?q=syntax&m=)	| 4th pos
runtime/cgo	| [search cgo](https://pkg.go.dev/search?q=cgo&m=)	| 2nd pos
runtime/metrics	| [search metrics](https://pkg.go.dev/search?q=metrics&m=)	| 21st pos
runtime/race	| [search race](https://pkg.go.dev/search?q=race&m=)	| 5th pos
runtime/trace	| [search trace](https://pkg.go.dev/search?q=trace&m=)	| 5th pos
testing/slogtest	| [search slogtest](https://pkg.go.dev/search?q=slogtest&m=)	| 2nd pos
text/template/parse	| [search parse](https://pkg.go.dev/search?q=parse&m=)	| 11th pos

**Comment by Deleplace** [comment]
I could clone go.googlesource.com/pkgsite, analyze the code, run the server locally and observe how the stdlib is not always prioritized.

However, I don't have the same datasources as production (postgres?) so a quick fix like "boost the Score of all stdlib results" might not be the best move.

**Comment by Deleplace** [comment]
cc @matloob @jamalc @tatianab

**Comment by komuw** [comment]
Related; 
1. https://github.com/golang/go/issues/40622 (builtin should be ranked higher in search)
2. https://github.com/golang/go/issues/41369 (standard library packages should be prioritized in search results)

**Comment by seankhliao** [comment]
Duplicate of #41369

**Comment by jba** [comment]
I agree with @hyangah, this is a separate issue from #41369 and I consider it a bug.

**Comment by jba** [comment]
Thanks @Deleplace for your careful investigation.

Currently pkgsite search does not implement the documented rule: "If the package path you specified is complete enough, matching a full package import path, you will be brought directly to the details page for the latest version of that package."

I'm not sure if that's really ideal, but this is actually a separate UI bug. As you can see below, the `path` package is mentioned when you search for "path".

![Image](https://github.com/user-attachments/assets/6649e9f7-f57d-48b3-9e16-13a8b21386c9)
It gets "demoted" because of a post-processing step we perform on searches. To prevent many packages in the same module from dominating the results, we keep only the highest-scoring package in the module at top level, and mention the rest as "related packages." `path/filepath` scores higher than `path` because it's more popular, so it stays at top level.

The fix for this particular problem is to avoid demoting stdlib packages.

Now does it make sense for a search of "path" to take you directly to pkg.go.dev/path? I say no. Then you could never search for general terms like "path", "sync" or "math". The rule may make sense for search terms with at least one `/`. What do people think of that revised rule?

**Comment by seankhliao** [comment]
I think the revised rule makes sense.

Related, can "std" be special cased to be the standard library? it would help since it doesn't have a common module prefix to match against.

**Comment by jba** [comment]
I mailed a CL to address the problem with "path" being hidden by "path/filepath". Now all stdlib packages will be at top level.

However, there may be more to do here:
- Why is image/color second to github.com/fatih/color despite being more popular?
-  Why is go/constant third after math and unicode/utf8?

I suspect these two have to do with the way we turn paths into tokens, which may be too delicate to tweak, but I'll look into that. (By "too delicate to tweak" I mean the amount of effort it would take to verify that changes to that algorithm do more good than harm exceed our work budget for pkgsite.)

Lastly, there is the rule that search terms with a slash that match exactly should visit the matched package directly, skipping search, with a special case for "std". That one probably needs its own issue.

**Comment by jba** [comment]
https://go.dev/cl/644676 will implement the "std/" tweak to search queries, and document that queries without slashes don't redirect, even if they match a package.

I looked into some of the other search concerns that @Deleplace mentioned in the top post. Here is a rundown on some of them:

- If you search for "constant", you see math and unicode/utf8 ahead of go/constant. That is because they are both far more popular than go/constant, and both mention "constant" in their package doc. Unfortunately, what they say is just "this is a package with some types and constants." Pkgsite's primitive search algorithm can't tell a sentence like that apart from one that says that the package is really _about_ constants. I don't see that improving any time soon.

- If you search for "cmp", you will get github.com/google/go-cmp/cmp before the stdlib cmp package, because it is slightly more popular and their synopses are similar. I expect that will change in the future.

- The maps package is now third, not eighth, when you search for "maps". It is beaten by golang.org/x/exp/maps, which is still slightly more popular, but that will change. The number 1 result is the very popular k8s.io/apimachinery/pkg/runtime. But that is more a flaw in our popularity algorithm than search, which is a topic for another day.
